storage/ctab/src/main/java/org/openscience/cdk/io/RdfileReader.java

```diff
@@ -0,0 +1,533 @@
+    static final String REACTION_EXT_REG = "$REREG";
+    static final String RXNFILE_START = "$RXN";
+    static final String MOLECULE_FMT = "$MFMT";
+    static final String MOLECULE_INT_REG = "$MIREG";
+    static final String MOLECULE_EXT_REG = "$MEREG";
+    static final String MOLFILE_START = "$MOL";
+    static final String M_END = "M  END";
+    static final String DTYPE = "$DTYPE";
+    static final String DATUM = "$DATUM";
+    static final String LINE_SEPARATOR_NEWLINE = "\n";
+    static final String PLUS_SIGN = "+";
+    private static final Pattern DTYPE_KEY = Pattern.compile("\\A\\" + DTYPE + " (.+)\\Z");
+    private static final Pattern DATUM_VALUE_FIRSTLINE = Pattern.compile("\\A\\" + DATUM + " (.{73}\\+|.{1,72})\\Z");
+    private static final Pattern MDL_CTAB_VERSION = Pattern.compile("(V[23]0{3})\\Z");
+    private static final Pattern MDL_RXN_VERSION = Pattern.compile("\\A\\$RXN ?(V30{3})?\\Z");
+
+    private boolean headerRead;
+    private String previousLine;
+    private int lineCounter;
+    private RdfileRecord nextRecord;
+    private boolean endOfFile;
+    private final boolean skipRecordsOnError;
+    private final BufferedReader bufferedReader;
+    private final IChemObjectBuilder chemObjectBuilder;
+
+    public RdfileReader(InputStream in, IChemObjectBuilder chemObjectBuilder) {
+        this(in, chemObjectBuilder, true);
+    }
+
+    public RdfileReader(InputStream in, IChemObjectBuilder chemObjectBuilder, boolean skipRecordsOnError) {
+        this(new InputStreamReader(in, StandardCharsets.UTF_8), chemObjectBuilder, skipRecordsOnError);
+    }
+
+    public RdfileReader(Reader reader, IChemObjectBuilder chemObjectBuilder) {
+        this(reader, chemObjectBuilder, true);
+    }
+
+    public RdfileReader(Reader reader, IChemObjectBuilder chemObjectBuilder, boolean skipRecordsOnError) {
+        this.bufferedReader = new BufferedReader(reader);
+        this.chemObjectBuilder = chemObjectBuilder;
+        this.skipRecordsOnError = skipRecordsOnError;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean hasNext() {
+        if (nextRecord != null)
+            return true;
+
+        if (endOfFile)
+            return false;
+
+        nextRecord = readNext();
+        return nextRecord != null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public RdfileRecord next() {
+        if (nextRecord == null && !endOfFile) {
+            nextRecord = readNext();
+        }
+
+        if (nextRecord == null) {
+            throw new NoSuchElementException("RdfileReader reached end of file.");
+        }
+
+        RdfileRecord returnValue = nextRecord;
+        nextRecord = null;
+        return returnValue;
+    }
+
+    RdfileRecord readNext() {
+        try {
+            return doReadNext();
+        } catch (IOException exception) {
+            LOGGER.error("I/O error when reading RDfile: " + exception.getMessage());
```

Comment:
What you can do here, is have a `STRICT` mode. When set true, you can throw an exception. This gives the user to ability to take action.

Second, you can use the IO listener functionality to report the error without failing, so that the user can record where in the file there are errors.

Reply:
(1) I do like the idea of the STRICT mode and might look into that.

(2) Can you elaborate on the IO listener functionality you are referring to? Is this an interface/base class in CDK?

Reply:
The IO listener stuff is how the options are handled. It's useful for UIs (e.g. Bioclipse) since you can list the options of a defaults automatically. In makes programmer/API usage very clunky though: 

For example, this is how you are "meant" to set the option "ForceReadAs3Dcoords" in V2000 reader. Note it's a String not a bool!:

```java
MDLV2000Reader reader = new MDLV2000Reader(ins);
Properties prop = new Properties();
prop.setProperty("ForceReadAs3DCoordinates", "true");
propertiesListener listener = new PropertiesListener(prop);
reader.addChemObjectIOListener(listener);
reader.customizeJob();
```

Personally I would have a gone a different way, but it is what we have. In some places I have added the directory access methods and utilities to make life a little smoother:

```java
SDFWriter sdfWriter = new SDFWriter(writer);
sdfWriter.getSetting(SDFWriter.OptAlwaysV3000).setSetting("true");
// or in this case I deem the option so important it has it's own method :-)
sdfWriter.setAlwaysV3000(true);
```

Reply:
Ah I miss-read Egon's input on the IO listener... personally I would prefer a return status on the record. So then the you force the user to decide what to do with them.

Reply:
Thanks for the very good explanation re the IO listener stuff. I saw that in the IO classes before but haven't paid too much attention.

However, if I understand correctly Egon was actually referring to something else? Are we talking about the IChemObject Interface and its handleError methods...? 🤷🏼

Reply:
Correct, it’s the error handler stuff.

Reply:
If I understand the hierarchy of the IO objects in CDK correctly that could then lead to the following class declaration

`public class RdfileReader extends DefaultIteratingChemObjectReader<IChemObject>`

if we want to stick with `IChemObject` as a return object for each record.
- I would pick up the error handling as implemented by `DefaultIteratingChemObjectReader`.
- The functionality of the boolean `skipRecordsOnError` would be substituted by `IChemObjectReader.Mode.RELAXED` and `IChemObjectReader.Mode.STRICT`
- `RdfileRecord` would just be an inner class of `RdfileReader`
- `CharIter` might end up us an inner class of `RdfileReader`, too; I only use a very limited set of methods there and would remove the unused methods
- and the external/internal registry number would be returned as a property of the molecule/reaction

Main drawback that I see: given the discussion with Daniel recently we cannot make this implementation thread-safe due to the `setReader(Reader)` and `setReader(InputStream)` methods that are part of the interface `IChemObjectReader`.

Does that make sense? Are we all happy with this?